javascript/frameworks/cap/lib/advanced_security/javascript/frameworks/cap/CAPCqlInjectionQuery.qll

```diff
@@ -0,0 +1,216 @@
+   */
+  abstract DataFlow::Node getQuery();
+}
+
+/**
+ * A CQL clause parameterized with a string concatentation expression.
+ */
+class CqlClauseWithStringConcatParameter instanceof CqlClause {
+  CqlClauseWithStringConcatParameter() {
+    exists(DataFlow::Node queryParameter |
+      (
+        if this instanceof CqlInsertClause or this instanceof CqlUpsertClause
+        then
+          queryParameter = this.getArgument().flow()
+          or
+          /*
+           * Account for cases where an object with a string concatenation is passed. e.g.
+           * ``` javascript
+           * let insertQuery = INSERT.into`SomeEntity`.entries({col1: "column_" + col});
+           * ```
+           */
+
+          queryParameter = this.getArgument().flow().(SourceNode).getAPropertyWrite().getRhs()
+        else queryParameter = this.getArgument().flow()
+      ) and
+      exists(StringConcatenation::getAnOperand(queryParameter))
+    )
+  }
+
+  Location getLocation() { result = super.getLocation() }
+
+  string toString() { result = super.toString() }
+}
+
+/**
+ * An await expression that has as its operand a CQL clause that includes a
+ * string concatenation operation.
+ */
+class AwaitCqlClauseWithStringConcatParameter extends CqlInjectionSink {
+  DataFlow::Node queryParameter;
+  DataFlow::Node query;
+  CqlClauseWithStringConcatParameter cqlClauseWithStringConcat;
+
+  AwaitCqlClauseWithStringConcatParameter() {
+    exists(AwaitExpr await |
+      this = await.flow() and
+      await.getOperand() = cqlClauseWithStringConcat.(CqlClause).asExpr()
+    )
+  }
+
+  override DataFlow::Node getQuery() { result = cqlClauseWithStringConcat.(CqlClause).flow() }
+}
+
+/**
+ * The first argument passed to the call to `cds.run`, `cds.db.run`, or `srv.run`
+ * whose value is a CQL query object that includes a string concatenation. e.g.
+ * ``` javascript
+ * // 1. CQN object constructed from Fluent API
+ * const query = SELECT.from`Entity1`.where("ID=" + id);
+ * cds.run(query);
+ *
+ * // 2. CQN object parsed from a string
+ * const query = cds.parse.cql("SELECT * from Entity1 where ID =" + id);
+ * cds.run(query);
+ *
+ * // 3. An unparsed CQL string (only valid in old versions of CAP)
+ * const query = "SELECT * from Entity1 where ID =" + id;
+ * Service2.run(query);
+ * ```
+ * The `getQuery/0` member predicate gets the `query` argument of the above calls
+ * to `run`.
+ */
+class StringConcatParameterOfCqlRunMethodQueryArgument extends CqlInjectionSink {
+  CqlRunMethodCall cqlRunMethodCall;
+
+  StringConcatParameterOfCqlRunMethodQueryArgument() {
+    this = cqlRunMethodCall.getAQueryParameter()
+  }
+
+  override DataFlow::Node getQuery() { result = this }
+}
+
+/**
+ * A CQL shortcut method call (`read`, `create`, ...) parameterized with a string
+ * concatenation expression. e.g.
+ * ``` javascript
+ * cds.read("Entity1").where(`ID=${id}`); // Notice the surrounding parentheses!
+ * cds.create("Entity1").entries({id: "" + id});
```

Comment:
I don't think this case is covered by this class.

Reply:
Oh, sorry for the confusion, it is already covered:
- [The CqlCreateClause class](https://github.com/advanced-security/codeql-sap-js/pull/200/files#diff-afddbc8d6075d4158b724713c1a12b83018634700e654c6739c513f8589dfcdaR925-R937).
- The test cases including the [CqlCreateClause](https://github.com/advanced-security/codeql-sap-js/pull/200/files#diff-e4c65e0ef9a88023ea50cc1f500260bedb6a06b3aaf18538a44db5a1cc7c1e6dR51-R68).

Reply:
Removed test cases where a call to `entries` is involved in [9101ff3](https://github.com/advanced-security/codeql-sap-js/pull/200/commits/9101ff31a4289d8c7d98a69f1bc42424689ef93f), and got rid of the related propagation step in [a763874](https://github.com/advanced-security/codeql-sap-js/pull/200/commits/a763874f95d6d4abae38eb7670da03e104f34b01).